2.0.0 - Overhaul mappings reader

jamierocks opened the pull request:
This commit completely replaces the existing code (I will bring back
the mappings joiner shortly), and makes the following improvements:

1. Using this library now feels more familar to using Lorenz.
2. This library now exposes more options for reading Tiny mappings:
     - Standard format
     - Legacy format
     - Detect format
3. The library now works for environments reading simple Tiny files,
   rather than just joined environments.
4. I have also bumped the Lorenz dependency to 0.5.2.

jamierocks committed:
2.0.0 - Overhaul mappings reader

This commit completely replaces the existing code (I will bring back
the mappings joiner shortly), and makes the following improvements:

1. Using this library now feels more familar to using Lorenz.
2. This library now exposes more options for reading Tiny mappings:
     - Standard format
     - Legacy format
     - Detect format
3. The library now works for environments reading simple Tiny files,
   rather than just joined environments.
4. I have also bumped the Lorenz dependency to 0.5.2.

Juuxel commented on src/main/java/net/fabricmc/lorenztiny/TinyMappingFormat.java:
```diff
@@ -0,0 +1,155 @@
+	 * @throws IOException if an I/O error occurs opening the file
+	 */
+	public MappingsReader createReader(final Path path,
+									   final String from, final String to) throws IOException, MappingParseException {
```
What's with the indentation here? (and some other places in this class and `TinyMappingsReader`)

jamierocks replied:
That's odd - must've been IntelliJ when I formatted my code (originally written in a none-tabs environment).

jamierocks replied:
So this wasn't actually an issue in my IDE - but is absolutely an issue with my IDE.

Basically it replaced the spaces (where they are used to match text) with tabs, causing it to look off (since GitHub has spaces shorter than 4 spaces) - I have replaced such tabs with spaces.

jamierocks committed:
Add mappings joiner

modmuss50 reviewed (approved):
Looks fine, I will intergrate with loom and merge soon.

modmuss50 commented on src/main/java/net/fabricmc/lorenztiny/TinyMappingsJoiner.java:
```diff
@@ -0,0 +1,120 @@
+			for (final FieldDef fieldB : classB.getFields()) {
+				final FieldDef fieldA = fields.get(fieldB.getName(this.matchB));
+
+				mapping.getOrCreateFieldMapping(fieldA.getName(this.from), fieldA.getDescriptor(this.from))
```
Im getting an NPE here:

```
Caused by: java.lang.NullPointerException: Cannot invoke "net.fabricmc.mapping.tree.FieldDef.getName(String)" because "fieldA" is null
	at net.fabricmc.lorenztiny.TinyMappingsJoiner.read(TinyMappingsJoiner.java:101)
	at org.cadixdev.lorenz.io.MappingsReader.read(MappingsReader.java:53)
	at net.fabricmc.loom.task.MigrateMappingsTask.migrateMappings(MigrateMappingsTask.java:153)
	at net.fabricmc.loom.task.MigrateMappingsTask.doTask(MigrateMappingsTask.java:100)
```

Seems to happen when the target mappigns have a field that the source mappings dont?

In loom I have:

```java
MappingSet mappingSet = new TinyMappingsJoiner(currentMappings, "named", targetMappings, "named", "intermediary").read();
```

The old code was:

```java
MappingSet mappingSet = LorenzTiny.readMappings(currentMappings, targetMappings, "intermediary", "named").read();
```

jamierocks committed:
Handle joins with missing mappings

This is closer to the original implementation, definately test this
thorougly though.

jamierocks committed:
Get method names properly

modmuss50 reviewed (approved):
Changes have been tested and it seems to work the same as before now.
